src/storage-client/src/types/sources/data.rs

```diff
@@ -0,0 +1,222 @@
+        }
+    }
+
+    fn from_proto(proto: ProtoSourceData) -> Result<Self, TryFromProtoError> {
+        use proto_source_data::Kind;
+        match proto.kind {
+            Some(kind) => match kind {
+                Kind::Ok(row) => Ok(SourceData(Ok(row.into_rust()?))),
+                Kind::Err(err) => Ok(SourceData(Err(err.into_rust()?))),
+            },
+            None => Result::Err(TryFromProtoError::missing_field("ProtoSourceData::kind")),
+        }
+    }
+}
+
+impl Codec for SourceData {
+    fn codec_name() -> String {
+        "protobuf[SourceData]".into()
+    }
+
+    fn encode<B: BufMut>(&self, buf: &mut B) {
+        self.into_proto()
+            .encode(buf)
+            .expect("no required fields means no initialization errors");
+    }
+
+    fn decode(buf: &[u8]) -> Result<Self, String> {
+        let proto = ProtoSourceData::decode(buf).map_err(|err| err.to_string())?;
+        proto.into_rust().map_err(|err| err.to_string())
+    }
+}
+
+impl From<Partitioned<PartitionId, MzOffset>> for SourceData {
+    fn from(partition: Partitioned<PartitionId, MzOffset>) -> SourceData {
+        use Datum::{Int32, Null};
+        use PartitionId::Kafka;
+        let mut row = Row::with_capacity(2);
+        let mut packer = row.packer();
+
+        match partition.interval() {
+            Interval::Range(l, u) => match (l, u) {
+                (RangeBound::Bottom, RangeBound::Top) => {
+                    packer
+                        .push_range(
+                            RangeBoundDesc::new(Null, false),
+                            RangeBoundDesc::new(Null, false),
+                        )
+                        .unwrap();
+                }
+                (RangeBound::Bottom, RangeBound::Elem(Kafka(pid))) => {
+                    packer
+                        .push_range(
+                            RangeBoundDesc::new(Null, false),
+                            RangeBoundDesc::new(Int32(*pid), false),
+                        )
+                        .unwrap();
+                }
+                (RangeBound::Elem(Kafka(pid)), RangeBound::Top) => {
+                    packer
+                        .push_range(
+                            RangeBoundDesc::new(Int32(*pid), false),
+                            RangeBoundDesc::new(Null, false),
+                        )
+                        .unwrap();
+                }
+                (RangeBound::Elem(Kafka(l_pid)), RangeBound::Elem(Kafka(u_pid))) => {
+                    packer
+                        .push_range(
+                            RangeBoundDesc::new(Int32(*l_pid), false),
+                            RangeBoundDesc::new(Int32(*u_pid), false),
+                        )
+                        .unwrap();
+                }
+                // PartitionId::None is not serialized to SourceData.
+                (RangeBound::Elem(PartitionId::None), _)
+                | (_, RangeBound::Elem(PartitionId::None)) => {}
```

Comment:
We should panic if we reach this point, right?

Reply:
Yep!